plugins/kafka/kafka_output.go

```diff
@@ -371,6 +373,53 @@ func (k *KafkaOutput) CleanupForRestart() {
 	return
 }
 
+// Escapes a topic name by encoding any non-valid kafka topic name
+func topicEscape(s string) string {
+	hexCount := 0
+	for i := 0; i < len(s); i++ {
+		if shouldEscape(s[i]) {
+			hexCount++
+		}
+	}
+
+	if hexCount == 0 {
+		return s
+	}
+
+	t := make([]byte, len(s)+2*hexCount)
+	j := 0
+	for i := 0; i < len(s); i++ {
+		c := s[i]
+		if shouldEscape(c) {
+			t[j] = '-'
+			t[j+1] = "0123456789ABCDEF"[c>>4]
+			t[j+2] = "0123456789ABCDEF"[c&15]
+			j += 3
+		} else {
+			t[j] = s[i]
+			j++
+		}
+	}
+	return string(t)
+}
+
+func shouldEscape(c byte) bool {
+	// List of valid kafka topic characters can be found at:
+	// https://github.com/apache/kafka/blob/43b92f8b1ce8140c432edf11b0c842f5fbe04120/core/src/main/scala/kafka/common/Topic.scala#L25
+	//   val legalChars = "[a-zA-Z0-9\\._\\-]"
+	if 'A' <= c && c <= 'Z' || 'a' <= c && c <= 'z' || '0' <= c && c <= '9' {
+		return false
+	}
+
+	switch c {
+	case '.', '_', '-':
```

Comment:
So there is one problem here, is the `-` character needs to be escaped, since that's being used as the escape character above on L394.

This will prevent backwards compatibility though if we escape `-`. So unsure how to proceed best without changing the topic names on people.